internal/cluster/cluster_test.go

```diff
@@ -4081,3 +4081,65 @@ var _ = Describe("ResetClusterFiles", func() {
 		Expect(capi.ResetClusterFiles(ctx, &cluster, mockObjectHandler)).To(BeNil())
 	})
 })
+
+var _ = Describe("update finalizing stage", func() {
+	var (
+		ctx       = context.Background()
+		db        *gorm.DB
+		dbName    string
+		clusterID strfmt.UUID
+		capi      API
+		ctrl      *gomock.Controller
+	)
+	BeforeEach(func() {
+		db, dbName = common.PrepareTestDB()
+		ctrl = gomock.NewController(GinkgoT())
+		capi = NewManager(getDefaultConfig(), common.GetTestLog(), db, commontesting.GetDummyNotificationStream(ctrl), nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, false)
+		clusterID = strfmt.UUID(uuid.New().String())
+	})
+	createCluster := func(status string) {
+		cls := common.Cluster{
+			Cluster: models.Cluster{
+				ID:     &clusterID,
+				Status: &status,
+				Progress: &models.ClusterProgressInfo{
+					FinalizingStageTimedOut: true,
+				},
+			},
+		}
+		Expect(db.Create(&cls).Error).ToNot(HaveOccurred())
+	}
+	AfterEach(func() {
+		common.DeleteTestDB(db, dbName)
+		ctrl.Finish()
+	})
+	It("cluster doesn't exist", func() {
+		err := capi.UpdateFinalizingStage(ctx, clusterID, models.FinalizingStageWaitingForFinalizing)
+		Expect(err).To(BeAssignableToTypeOf(&common.ApiErrorResponse{}))
+		Expect(err.(*common.ApiErrorResponse).Code()).To(BeEquivalentTo(http.StatusNotFound))
+	})
+	It("cluster in wrong status", func() {
+		createCluster(models.ClusterStatusReady)
+		err := capi.UpdateFinalizingStage(ctx, clusterID, models.FinalizingStageWaitingForFinalizing)
+		Expect(err).To(BeAssignableToTypeOf(&common.ApiErrorResponse{}))
+		Expect(err.(*common.ApiErrorResponse).Code()).To(BeEquivalentTo(http.StatusBadRequest))
+	})
+	It("unexpected stage", func() {
+		createCluster(models.ClusterStatusFinalizing)
+		err := capi.UpdateFinalizingStage(ctx, clusterID, "unexpected stage")
+		Expect(err).To(BeAssignableToTypeOf(&common.ApiErrorResponse{}))
+		Expect(err.(*common.ApiErrorResponse).Code()).To(BeEquivalentTo(http.StatusBadRequest))
+	})
+	for _, st := range finalizingStages {
+		stage := st
+		It(fmt.Sprintf("success for stage '%s'", stage), func() {
+			createCluster(models.ClusterStatusFinalizing)
+			Expect(capi.UpdateFinalizingStage(ctx, clusterID, stage)).ToNot(HaveOccurred())
+			cls, err := common.GetClusterFromDB(db, clusterID, common.SkipEagerLoading)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(cls.Progress.FinalizingStage).To(Equal(stage))
+			Expect(time.Time(cls.Progress.FinalizingStageStartedAt)).To(BeTemporally("~", time.Now(), time.Second))
```

Comment:
To make this less fragile you could pass the time to `UpdateFinalizingStage`.

Reply:
It doesn't look nice.  But I can modify the test to be less fragile.